src/writer/stripe.rs

```diff
@@ -0,0 +1,175 @@
+    /// to required batch size.
+    pub fn encode_batch(&mut self, batch: &RecordBatch) -> Result<()> {
+        // TODO: consider how to handle nested types (including parent nullability)
+        for (array, encoder) in batch.columns().iter().zip(self.columns.iter_mut()) {
+            encoder.encode_array(array)?;
+        }
+        self.row_count += batch.num_rows();
+        Ok(())
+    }
+
+    /// Flush streams to the writer, and write the stripe footer to finish
+    /// the stripe. After this, the [`StripeWriter`] will be reset and ready
+    /// to write a fresh new stripe.
+    pub fn finish_stripe(&mut self) -> Result<StripeInformation> {
+        let start_offset = self.writer.stream_position().context(IoSnafu)?;
+
+        // Order of column_encodings needs to match final type vec order.
+        // (see arrow_writer::serialize_schema())
+        // Direct encoding to represent root struct
+        let mut column_encodings = vec![ColumnEncoding::Direct];
+        let child_column_encodings = self
+            .columns
+            .iter()
+            .map(|c| c.column_encoding())
+            .collect::<Vec<_>>();
+        column_encodings.extend(child_column_encodings);
+        let column_encodings = column_encodings.iter().map(From::from).collect();
+
+        // Root type won't have any streams
+        let mut written_streams = vec![];
+        for (index, c) in self.columns.iter_mut().enumerate() {
+            // Offset by 1 to account for root of 0
+            let column = index + 1;
+            let streams = c.finish();
+            // Flush the streams to the writer
+            for s in streams {
+                let (s_type, bytes) = s.into_parts();
+                let length = bytes.len();
+                self.writer.write_all(&bytes).context(IoSnafu)?;
+                written_streams.push(WrittenStream {
+                    s_type,
+                    column,
+                    length,
+                });
+            }
+        }
+        let streams = written_streams.iter().map(From::from).collect();
+        let stripe_footer = proto::StripeFooter {
+            streams,
+            columns: column_encodings,
+            writer_timezone: None,
+            encryption: vec![],
+        };
+
+        let data_length = self.writer.stream_position().context(IoSnafu)? - start_offset;
+        let footer_bytes = stripe_footer.encode_to_vec();
+        let footer_length = footer_bytes.len() as u64;
+        let row_count = self.row_count;
+        self.writer.write_all(&footer_bytes).context(IoSnafu)?;
+
+        // Reset state for next stripe
+        self.row_count = 0;
+
+        Ok(StripeInformation {
+            start_offset,
+            index_length: 0,
+            data_length,
+            footer_length,
+            row_count,
+        })
+    }
+
+    /// Close this writer and return the inner writer.
+    pub fn close(self) -> W {
```

Comment:
Calling `close` to return the internal `W` is a bit surprising. How about using `finish` instead?

Reply:
Thanks, fixed